cargo-marker/src/backend/cargo.rs

```diff
@@ -0,0 +1,78 @@
+use std::process::Command;
+
+use camino::Utf8PathBuf;
+use cargo_metadata::MetadataCommand;
+use serde::Deserialize;
+
+use crate::ExitStatus;
+
+#[derive(Debug, Default)]
+pub struct Cargo {
+    /// The rustc toolchain this driver belongs to. This can be `None` during
+    /// execution commands such as `cargo locate-project`
+    pub(crate) toolchain: Option<String>,
+}
+
+#[derive(Deserialize, Debug)]
+struct ProjectLocation {
+    root: Utf8PathBuf,
+}
+
+impl Cargo {
+    pub fn with_toolchain(toolchain: impl Into<String>) -> Self {
+        Self {
+            toolchain: Some(toolchain.into()),
+        }
+    }
+
+    /// This returns a command calling rustup, which acts as a proxy for the
+    /// Cargo of the selected toolchain.
+    /// It may add additional flags for verbose output.
+    ///
+    /// See also [`super::toolchain::Toolchain::cargo_build_command`] if the
+    /// commands is intended to build a crate.
+    pub fn command(&self) -> Command {
+        // Marker requires rustc's shared libraries to be available. These are
+        // added by rustup, when it acts like a proxy for cargo/rustc/etc.
+        // This also means that cargo needs to be invoked via rustup, to have
+        // the libraries available when Marker is invoked. This is such a mess...
+        // All of this would be so, so much simpler if marker was part of rustup :/
+        if let Some(toolchain) = &self.toolchain {
+            let mut cmd = Command::new("rustup");
+            cmd.args(["run", toolchain, "cargo"]);
+            cmd
+        } else {
+            // for cargo locate-project - this command can be run without
+            // specified toolchain
+            Command::new("cargo")
+        }
+    }
+
+    pub fn cargo_locate_project(&self) -> Result<Utf8PathBuf, ExitStatus> {
+        let mut cmd = self.command();
+
+        let output = cmd
+            .arg("locate-project")
+            .arg("--workspace")
+            .output()
+            .map_err(|err| ExitStatus::fatal(err, "error locating workspace manifest Cargo.toml"))?;
+
+        if !output.status.success() {
+            let stderr = String::from_utf8_lossy(&output.stderr);
+            return Err(ExitStatus::fatal(
+                stderr.trim(),
+                format!("cargo locate-project failed with {}", output.status),
+            ));
+        }
+
+        let manifest_location: ProjectLocation = serde_json::from_slice(&output.stdout)
+            .map_err(|err| ExitStatus::fatal(err, "failed to deserialize cargo locate-project output"))?;
+
+        Ok(manifest_location.root)
+    }
+
+    pub fn metadata(&self) -> MetadataCommand {
+        let _ = self;
+        MetadataCommand::new()
```

Comment:
As @Veetaha mentioned, this problem exists only on Windows

Yes, I do it on purpose. Clippy requires `self` to be used inside the function body, otherwise it warns about the following [lint](https://rust-lang.github.io/rust-clippy/master/index.html#unused_self). 
This is the second possible solution to the first problem. A `cargo metadata` command is created, which takes path to `cargo` in `Path`.

And here path to `cargo` of another version of toolchain is set and the drift of versions occurs

https://github.com/rust-marker/marker/blob/ef23c8b2aff12ae7bb8f44ef8a911e94d225cef3/cargo-marker/src/backend/toolchain.rs#L83-L86

I used `CargoMetadata`, because it's implements adding arguments, executing a command, handling errors, and parsing the result... So it was easier for me to use it :)

Reply:
> Clippy requires self to be used inside the function body, otherwise it warns about the following lint.

Can you change this into using `#[allow(clippy::unused_self)]` instead?

> This is the second possible solution to the first problem. A `cargo metadata` command is created, with path to cargo resolved from `Path`.

And: @Veetaha:

> As for the metadata command, I think it is reasonable to use whatever the toolchain is installed in the user's workspace to read the metadata. We also don't want to accidentally autoinstall the nightly-2023-07-13 when running the cargo metadata command.

The reason, why the previous fix also set the `RUSTUP_TOOLCHAIN` for the metadata command, was to make sure that we know what we can expect. #188 was caused by reading a metadata field, which was not yet available in the workspace version of Rust.

So this question is about, not reintroducing #188 or other errors which might happen if an old toolchain version is used.

Reply:
@xFrednet as I see #188 was caused exactly by the version drift between cargo and rustc as was described in this PR and in the comment https://github.com/rust-marker/marker/issues/188#issuecomment-1652848485.

This version drift was *supposedly* fixed by adding `RUST_TOOLCHAIN=nightly-2023-07-13` to cargo metadata invocation, however it fixed things on linux only, because we know only linux that we invoke `cargo_path` pointing to real cargo `nightly-2023-07-13` version of `cargo`, and `PATH` is not modified.

However, on windows `rustup` modifies `Path` adding the path to the toolchain bin dir, so this version drift happens. It's hard to explain it in text, that's why there are diagrams in the PR description.

In conclusion
## Before
- Linux (no drift happens here, when `cargo` invokes `rustc`, they are both of `nightly-2023-07-13` version):
   - `cargo_path`: `~/.rustup/toolchains/nightly-2023-07-13/bin/cargo`
   - `PATH`: `~/.cargo/bin`
   - `RUSTUP_TOOLCHAIN`: `nightly-2023-07-13`
- Windows (drift happens, `cargo` of version `nightly-2023-07-13` invokes `rustc` of version `1.65.0`)
   - `cargo_path`: `~/.rustup/toolchains/nightly-2023-07-13/cargo`
   - `Path`: `~/.rustup/toolchains/nightly-2023-07-13/bin;~/.cargo/bin`
   - `RUSTUP_TOOLCHAIN`: `nightly-2023-07-13` (this gets ignored, because `Path` will make cargo call rustc directly, not via rustup proxy)

## Now
- Linux (no drift happens here, when `cargo` invokes `rustc`, they are both of some `X` version):
   - `PATH`: `~/.cargo/bin`
- Windows (no drift happens here, when `cargo` invokes `rustc`, they are both of some `X` version):
   - `Path`: `~/.rustup/toolchains/X/bin;~/.cargo/bin`

Meaning there is now consistent behaviour between linux and windows.

So there is no such problem as you described nether before, nor now:
>  reading a metadata field, which was not yet available in the workspace version of Rust.

The `cargo_metadata` crate is completely backwards compatible, because it uses `--format-version 1` to make sure it works across all versions of `cargo`. So there is no need to run `cargo metadata` with a specific version oc `cargo`. Let's not do that.